src/main/java/com/mongodb/kafka/connect/sink/writemodel/strategy/DeleteOneBusinessKeyStrategy.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.kafka.connect.sink.writemodel.strategy;
+
+import static com.mongodb.kafka.connect.sink.MongoSinkTopicConfig.ID_FIELD;
+import static com.mongodb.kafka.connect.sink.writemodel.strategy.WriteModelHelper.flattenKeys;
+
+import org.apache.kafka.connect.errors.DataException;
+
+import org.bson.BSONException;
+import org.bson.BsonDocument;
+
+import com.mongodb.client.model.DeleteOneModel;
+import com.mongodb.client.model.WriteModel;
+
+import com.mongodb.kafka.connect.sink.Configurable;
+import com.mongodb.kafka.connect.sink.MongoSinkTopicConfig;
+import com.mongodb.kafka.connect.sink.converter.SinkDocument;
+import com.mongodb.kafka.connect.sink.processor.id.strategy.IdStrategy;
+import com.mongodb.kafka.connect.sink.processor.id.strategy.PartialKeyStrategy;
+import com.mongodb.kafka.connect.sink.processor.id.strategy.PartialValueStrategy;
+
+public class DeleteOneBusinessKeyStrategy implements WriteModelStrategy, Configurable {
+
+  private boolean isPartialId = false;
+
+  @Override
+  public WriteModel<BsonDocument> createWriteModel(final SinkDocument document) {
+    BsonDocument vd =
+        document
+            .getValueDoc()
+            .orElseThrow(
+                () ->
+                    new DataException(
+                        "Could not build the WriteModel,the value document was missing unexpectedly"));
+
+    try {
+      BsonDocument businessKey = vd.getDocument(ID_FIELD);
+      vd.remove(ID_FIELD);
+      if (isPartialId) {
+        businessKey = flattenKeys(businessKey);
+      }
+      return new DeleteOneModel<>(businessKey);
+
+    } catch (BSONException e) {
+      throw new DataException(
+          "Could not build the WriteModel,the value document does not contain an _id field of"
+              + " type BsonDocument which holds the business key fields.`.");
+    }
+  }
+
+  @Override
+  public void configure(final MongoSinkTopicConfig configuration) {
+    IdStrategy idStrategy = configuration.getIdStrategy();
+    isPartialId =
+        idStrategy instanceof PartialKeyStrategy || idStrategy instanceof PartialValueStrategy;
```

Comment:
I am blatantly unfamiliar with the semantics of `IdStrategy`, `WriteModelStrategy`, and do not know whether or not users are allowed to create their own implementations of `IdStrategy`. But if I were to judge based only on this code, I would say that it suggests the need for either a new method `IdStrategy.isPartial`, or a marker interface `IdStrategy.Partial`, or a similar mechanism that does not require modifying this code when a new partial `IdStrategy` implementation is born.

Reply:
That would be possible, if ever a new partial strategy is added then that would be a good time to add it.

Reply:
What concerns me in this approach is that the future you creating a new partial `IdStrategy` will have to remember that this place in the codebase needs to be updated. And if it's not you, but someone else, how will that someone know that this place must also be updated?

If we had [sealed classes](https://openjdk.java.net/jeps/360), we could have left a comment near the set of all allowed subclasses about updating this place in code. Not a good solution, but at least something. However, we do not have that language feature yet (and will not have it in the driver code for the foreseeable future).

Reply:
Thats still fine, its a limitation of the current implementations.Users can create their own implementations of both WriteModelStrategy (where this initial code came from) and IdStrategies. 

It hasn't been an issue for the alternative BusinessKeyStrategies which have existed for the years of this project.

Reply:
Is my understanding correct that the current argument goes like the following?
1. Other code in the codebase, e.g., `ReplaceOneBusinessKeyStrategy` has the same (objective) issue.
2. It has been fine so far.
3. Therefore, we may continue writing code that has the issue.

If yes, then I can neither consider this as a good argument nor agree with it because
- the change needed to avoid the issue in the code is neither disruptive nor too laborious (if I am wrong about this part, please tell me);
- the fact that the same mechanism of differentiating partial `IdStrategy` is needed in other places in the codebase emphasizes the need to have this mechanism instead of repeating the same problematic workaround.

I don't know how we approach situations when the author and a reviewer cannot come to an agreement in a timely manner. Here is what I propose (again, provided that my understanding of your current argument I described above is correct):
- create ("split from" KAFKA-168) a `KAFKA` improvement about introducing a mechanism of differentiating partial `IdStrategy`;
- approve and merge the current PR.

Reply:
Created [KAFKA-199 Introduce a mechanism of differentiating partial `IdStrategy`](https://jira.mongodb.org/browse/KAFKA-199).